packages/client/src/runtime/DataProxyEngine.ts

```diff
@@ -0,0 +1,243 @@
+      }
+
+      if (!res.ok) {
+        throw new Error('GraphQL request failed')
+      }
+
+      return JSON.parse(await res.text())
+    } catch (err) {
+      if (n >= MAX_RETRIES) {
+        this.logEmitter.emit('error', `Failed to query: ${err.message}`)
+        throw err
+      } else {
+        const delay = await backOff(n)
+        this.logEmitter.emit('warn', `Retrying after ${delay}`)
+        return this.request<T>(q, {}, n + 1)
+      }
+    }
+  }
+
+  requestBatch<T>(qs: string[], _hs: any, transact = false) {
+    // TODO: transactions not supported (see below)
+    // TODO: introduce batch endpoint in Data Proxy;
+    // current implementation will result in `qs.length` HTTP calls
+    return transact
+      ? this.transaction()
+      : Promise.all(qs.map((q) => this.request<T>(q)))
+  }
+
+  // TODO: figure out how to support transactions
+  transaction(): Promise<any> {
+    throw new Error('Transactions are currently not supported in Data Proxy')
+  }
+}
+
+// Regex + prismafile feels fragile...
+// TODO: build a minimal single-purpose (URL extraction)
+// and well-tested TS schema parser
+function extractHostAndApiKey(schemaText: string) {
+  // Extract the datasource block so we feed prismafile with minimal input
+  const strippedToDatasource = /datasource[ \t]+[^\s]+[ \t]+\{[^}]+}/.exec(
+    schemaText,
+  )
+  if (!strippedToDatasource) {
+    throw new Error('Could not find a valid datasource block in the schema')
+  }
+
+  const datasourceText = strippedToDatasource[0]
+  if (typeof datasourceText !== 'string' || datasourceText.length < 1) {
+    throw new Error('Could not find a valid datasource block in the schema')
+  }
+
+  // Use prismafile to determine the URL
+  let url = ''
+  try {
+    const schemaAst = prismafile.parse(datasourceText)
+    const [datasource] = schemaAst.blocks.filter(
+      (b) => b.type === 'datasource',
+    ) as any[]
+
+    // Datasource type is not exported from prismafile, so from now on
+    // we're in any-land. Hence the dirtiness and the big try/catch around.
+
+    const urlAssignment = datasource.assignments.find(
+      (a: any) => a.key?.type === 'identifier' && a.key?.name === 'url',
+    )
+
+    if (!urlAssignment || !urlAssignment.value) {
+      throw new Error('No assignment found')
+    }
+
+    const { type, value, name, arguments: args } = urlAssignment.value
+
+    if (type === 'string_value' && typeof value === 'string') {
+      url = value // Datasource URL directly in the schema
+    } else if (
+      type === 'function_value' &&
+      name?.type === 'identifier' &&
+      name.name === 'env' &&
+      Array.isArray(args)
+    ) {
+      const arg = args[0]
+
+      if (arg?.type === 'string_value' && typeof arg.value === 'string') {
+        const envVarName = arg.value.trim()
+
+        if (envVarName.length > 0) {
+          // Datasource URL in environment
+          url = process.env[envVarName] ?? ''
+        }
```

Comment:
I think here reading the env var from the `.env` file might be needed as well.

It's actually passed to `_engineConfig` see
https://github.com/prisma/prisma/blob/main/packages/client/src/runtime/getPrismaClient.ts#L396:L396